server/src/main/java/org/opensearch/action/admin/indices/segments/TransportPitSegmentsAction.java

```diff
@@ -0,0 +1,259 @@
+    private final IndicesService indicesService;
+    private final SearchService searchService;
+    private final NamedWriteableRegistry namedWriteableRegistry;
+    private final TransportService transportService;
+    private final PitService pitService;
+
+    @Inject
+    public TransportPitSegmentsAction(
+        ClusterService clusterService,
+        TransportService transportService,
+        IndicesService indicesService,
+        ActionFilters actionFilters,
+        IndexNameExpressionResolver indexNameExpressionResolver,
+        SearchService searchService,
+        NamedWriteableRegistry namedWriteableRegistry,
+        PitService pitService
+    ) {
+        super(
+            PitSegmentsAction.NAME,
+            clusterService,
+            transportService,
+            actionFilters,
+            indexNameExpressionResolver,
+            PitSegmentsRequest::new,
+            ThreadPool.Names.MANAGEMENT
+        );
+        this.clusterService = clusterService;
+        this.indicesService = indicesService;
+        this.searchService = searchService;
+        this.namedWriteableRegistry = namedWriteableRegistry;
+        this.transportService = transportService;
+        this.pitService = pitService;
+    }
+
+    /**
+     * Execute PIT segments flow for all PITs or request PIT IDs
+     */
+    @Override
+    protected void doExecute(Task task, PitSegmentsRequest request, ActionListener<IndicesSegmentResponse> listener) {
+        if (request.getPitIds().isEmpty()) {
+            pitService.getAllPits(ActionListener.wrap(response -> {
+                request.setPitIds(response.getPitInfos().stream().map(ListPitInfo::getPitId).collect(Collectors.toList()));
+                getDoExecute(task, request, listener);
+            }, listener::onFailure));
+        } else {
+            getDoExecute(task, request, listener);
+        }
+    }
+
+    private void getDoExecute(Task task, PitSegmentsRequest request, ActionListener<IndicesSegmentResponse> listener) {
+        super.doExecute(task, request, listener);
+    }
+
+    /**
+     * This adds list of shards on which we need to retrieve pit segments details
+     * @param clusterState    the cluster state
+     * @param request         the underlying request
+     * @param concreteIndices the concrete indices on which to execute the operation
+     */
+    @Override
+    protected ShardsIterator shards(ClusterState clusterState, PitSegmentsRequest request, String[] concreteIndices) {
+        final ArrayList<ShardRouting> iterators = new ArrayList<>();
+        for (String pitId : request.getPitIds()) {
+            SearchContextId searchContext = decode(namedWriteableRegistry, pitId);
+            for (Map.Entry<ShardId, SearchContextIdForNode> entry : searchContext.shards().entrySet()) {
+                final SearchContextIdForNode perNode = entry.getValue();
+                if (Strings.isEmpty(perNode.getClusterAlias())) {
```

Comment:
Any reason for ClusterAlias to be empty for getting the list of Shards?

Reply:
Only for local cluster we are making this call.

Reply:
@bharath-techie could you please add this comment into the code? thank you